tracing/src/main/java/org/jboss/resteasy/tracing/api/RESTEasyTracingInfo.java

```diff
@@ -1,69 +1,94 @@
 package org.jboss.resteasy.tracing.api;
 
-import java.util.ArrayList;
-import java.util.List;
+import org.jboss.resteasy.tracing.api.providers.TextBasedRESTEasyTracingInfo;
 
-public class RESTEasyTracingInfo {
+import java.util.Iterator;
+import java.util.ServiceLoader;
 
-    public enum FORMAT {
-        TEXT,
-        JSON
-    }
+public abstract class RESTEasyTracingInfo implements RESTEasyTracingInfoInterface {
 
-    protected final List<RESTEasyTracingMessage> messageList = new ArrayList<RESTEasyTracingMessage>();
+    protected static final ServiceLoader<RESTEasyTracingInfo> INSTANCES;
+    protected static final RESTEasyTracingInfo DEFAULT = new TextBasedRESTEasyTracingInfo();
 
-    public void addMessage(RESTEasyTracingMessage message) {
-        messageList.add(message);
+    static {
+        INSTANCES = ServiceLoader.load(RESTEasyTracingInfo.class, Thread.currentThread().getContextClassLoader());
     }
 
-    public static String formatDuration(final long duration) {
-        if (duration == 0) {
-            return " ----";
+    public static RESTEasyTracingInfo get(String format) {
+        if (format == null || format.isEmpty()) {
+            return DEFAULT;
         } else {
-            return String.format("%5.2f", (duration / 1000000.0));
+            if (format.equals(RESTEasyTracingInfoFormat.TEXT.toString())) {
+                return new TextBasedRESTEasyTracingInfo();
+            } else if (format.equals(RESTEasyTracingInfoFormat.JSON.toString())) {
+                Iterator<RESTEasyTracingInfo> iter = INSTANCES.iterator();
+                while(iter.hasNext()) {
+                    RESTEasyTracingInfo instance = iter.next();
+                    if (containsIgnoreCase(instance.getClass().getCanonicalName(), "json")) {
+                        return instance;
+                    }
+                }
+            }
         }
+        return DEFAULT;
     }
 
-    public static String formatDuration(final long fromTimestamp, final long toTimestamp) {
-        return formatDuration(toTimestamp - fromTimestamp);
+    private static boolean containsIgnoreCase(final CharSequence str, final CharSequence searchStr) {
+        if (str == null || searchStr == null) {
+            return false;
+        }
+        final int len = searchStr.length();
+        final int max = str.length() - len;
+        for (int i = 0; i <= max; i++) {
+            if (regionMatches(str, true, i, searchStr, 0, len)) {
+                return true;
+            }
+        }
+        return false;
     }
 
-    public static String formatPercent(final long value, final long top) {
-        if (value == 0) {
-            return "  ----";
-        } else {
-            return String.format("%6.2f", 100.0 * value / top);
+    private static boolean regionMatches(final CharSequence cs, final boolean ignoreCase, final int thisStart,
+                                 final CharSequence substring, final int start, final int length)    {
```

Comment:
mmh... I don't like this check on the name that much, tbh. Why don't you add a method to the RESTEasyTracingInfo to check if a given instance supports a RESTEasyTracingFormat provided as parameter to such method? That way you could simply call the method in the loop over the RESTEasyTracingInfo returned by the service loader.

Reply:
Yes sir! Let me fix this.